lib/phoenix_live_view/colocated_js.ex

```diff
@@ -0,0 +1,337 @@
+      script does not actually export anything.
+
+    * `key` - A key under which to namespace the export. This is used by `Phoenix.LiveView.ColocatedHook` to
+      nest all hooks under the `hooks` key. For example, you could set this to `web_components` for each colocated
+      script that defines a web component and then access all of them as `colocated.web_components` when importing
+      the manifest.
+
+    * `extension` - a custom extension to use when writing the extracted file. The default is `js`.
+
+    * `manifest` - a custom manifest file to use instead of the default `index.js`. For example,
+      `web_components.ts`.
+
+  """
+
+  @behaviour Phoenix.Component.MacroComponent
+
+  @impl true
+  def transform({"script", attributes, [text_content]} = _ast, meta) do
+    opts = Map.new(attributes)
+    validate_name!(opts)
+    data = extract(opts, text_content, meta)
+
+    # we always drop colocated JS from the rendered output
+    {:ok, "", data}
+  end
+
+  def transform(_ast, _meta) do
+    raise ArgumentError, "ColocatedJS can only be used on script tags"
+  end
+
+  defp validate_name!(opts) do
+    case opts do
+      %{"name" => name} when is_binary(name) ->
+        :ok
+
+      %{"name" => name} ->
+        raise ArgumentError,
+              "the name attribute of a colocated JS / hook must be a compile-time string. Got: #{inspect(name)}"
+
+      _ ->
+        raise ArgumentError, "missing required name attribute for ColocatedJS"
+    end
+  end
+
+  @doc false
+  def extract(opts, text_content, meta) do
+    # _build/dev/phoenix-colocated/otp_app/MyApp.MyComponent/line_no.js
+    target_path =
+      target_dir()
+      |> Path.join(to_string(current_otp_app()))
+      |> Path.join(inspect(meta.env.module))
+
+    filename_opts =
+      %{name: opts["name"]}
+      |> maybe_put_opt(opts, "key", :key)
+      |> maybe_put_opt(opts, "manifest", :manifest)
+
+    filename = "#{meta.line}.#{opts["extension"] || "js"}"
+
+    File.mkdir_p!(target_path)
+    File.write!(Path.join(target_path, filename), text_content)
+
+    {filename, filename_opts}
+  end
+
+  defp maybe_put_opt(map, opts, opts_key, target_key) do
+    case opts do
+      %{^opts_key => value} ->
+        Map.put(map, target_key, value)
+
+      _ ->
+        map
+    end
+  end
+
+  @doc false
+  def compile do
+    # this step runs after all modules have been compiled
+    # so we can write the final manifests and remove outdated hooks
+    target_dir = target_dir()
+    clear_manifests!(target_dir)
+    files = clear_outdated_and_get_files!(target_dir)
+    write_new_manifests!(target_dir, files)
+  end
+
+  defp clear_manifests!(target_dir) do
+    manifests =
+      Path.wildcard(Path.join(target_dir, "*/*.*"))
+      |> Enum.filter(&File.regular?(&1))
+
+    for manifest <- manifests, do: File.rm!(manifest)
+  end
+
+  defp clear_outdated_and_get_files!(target_dir) do
+    apps = subdirectories(target_dir)
+
+    Enum.flat_map(apps, fn app_dir ->
```

Comment:
I think we should only traverse the current app. Each app is responsible to leave its folder in a workable state, no?

Reply:
oh I was somehow assuming this would only run once, but I guess the compiler would run for every project that specifies it, so it there are 10 dependencies with colocated js / hooks and also 2 Phoenix apps in an umbrella, it would run for each one separately, so that makes sense :)